synapse/storage/devices.py

```diff
@@ -557,6 +557,9 @@ def _prune_txn(txn):
             txn.execute(select_sql)
             rows = txn.fetchall()
 
+            if not rows:
+                return
```

Comment:
If you throw in a `count` then you could probably discard all the `destination, user_id` pairs that only have a single row in them.